### PR TITLE
[SYSTEMML-1626] x = solve(A,b) should report an error when it is an underdetermined system (in CP mode)

### PR DESCRIPTION
**Pull request by j143**
**What is the change?**
The hop (BinaryOp), throws an error - if the A matrix in solve(A, b) do not have squared dimensions.

### CODE REVIEW

**Comment by j143** [comment]
Test this Jenkins

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2342/<h2>Failed Tests: <span class='status-failure'>2</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2342/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>2</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2342/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix_full_other/FullMatrixMultiplicationUltraSparseTest/testMMUltraSparseUltraSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix_full_other.FullMatrixMultiplicationUltraSparseTest.testMMUltraSparseUltraSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2342/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix_full_other/FullMatrixMultiplicationUltraSparseTest/testMMUltraSparseUltraSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix_full_other.FullMatrixMultiplicationUltraSparseTest.testMMUltraSparseUltraSparseMR</strong></a></li></ul>

File: src/main/java/org/apache/sysml/hops/BinaryOp.java [hunk 1]
```diff
@@ -920,6 +920,8 @@ else if( op == OpOp2.RBIND ) {
 				return new long[]{ldim1, ldim2, lnnz};
 		}
 		else if ( op == OpOp2.SOLVE ) {
+			if (mc[0].getRows() != mc[0].getCols())
+				throw new RuntimeException("The A matrix, in solve(A,b) should have squared dimensions.");
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/hops/BinaryOp.java]: Could you move this check please into (1) the builtin expression validation, and (2) the runtime operation to ensure the error is always thrown independent of the size propagation being able to infer the size or not. Also not that this method for inferring worst-case output sizes is only executed if we cannot exactly infer the output sizes.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2345/

**Comment by mboehm7** [comment]
LGTM - thanks @j143

**Pull request closed without merging**